extensions/amp-stream-gallery/1.0/stream-gallery.js

```diff
@@ -0,0 +1,221 @@
+    />
+  );
+
+  return (
+    <ContainWrapper
+      layout={true}
+      size={true}
+      wrapperStyle={style}
+      contentStyle={{display: 'flex'}}
+      contentRef={galleryRef}
+      {...rest}
+    >
+      <CarouselContext.Provider value={carouselContext}>
+        {outsetArrows && arrowPrev}
+        <BaseCarousel
+          ref={carouselRef}
+          advanceCount={Math.floor(visibleCount)}
+          arrowPrev={arrowPrev}
+          arrowNext={arrowNext}
+          controls={outsetArrows ? Controls.NEVER : insetArrowVisibility}
+          loop={loop}
+          outsetArrows={outsetArrows}
+          snap={snap}
+          snapAlign={slideAlign}
+          style={{
+            flexGrow: 1,
+            maxWidth: maxContainerWidth,
+            justifyContent: extraSpace === 'around' ? 'center' : 'initial',
+          }}
+          visibleCount={visibleCount}
+        >
+          {children}
+        </BaseCarousel>
+        {outsetArrows && arrowNext}
+      </CarouselContext.Provider>
+    </ContainWrapper>
+  );
+}
+
+/**
+ * @param {!StreamGalleryDef.ArrowProps} props
+ * @return {PreactDef.Renderable}
+ */
+function DefaultArrow({advance, by, outsetArrows, ...rest}) {
+  const classes = useStyles();
+  return (
+    <button
+      onClick={() => advance(by)}
+      className={`${classes.arrow} ${
+        by < 0 ? classes.arrowPrev : classes.arrowNext
+      } ${outsetArrows ? classes.outsetArrow : classes.insetArrow}`}
+      aria-hidden="true"
+      {...rest}
+    >
+      {by < 0 ? (
+        <svg viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg">
+          <path
+            d="M14,7.4 L9.4,12 L14,16.6"
+            fill="none"
+            stroke="#000"
+            stroke-width="2"
+            stroke-linejoin="round"
+            stroke-linecap="round"
+          />
+        </svg>
+      ) : (
+        <svg viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg">
+          <path
+            d="M10,7.4 L14.6,12 L10,16.6"
+            fill="none"
+            stroke="#000"
+            stroke-width="2"
+            stroke-linejoin="round"
+            stroke-linecap="round"
+          />
+        </svg>
+      )}
+    </button>
+  );
+}
+
+/**
+ * Updates the number of items visible for the internal carousel based on
+ * the min/max item widths and how much space is available.
+ * @param {number} maxItemWidth
+ * @param {number} minItemWidth
+ * @param {number} maxVisibleCount
+ * @param {number} minVisibleCount
+ * @param {number} slideCount
+ * @param {number} peek
+ * @param {{current: Element}} galleryRef
+ * @return {{visibleCount: number, maxContainerWidth: number}}
+ */
+function getVisibleCount(
+  maxItemWidth,
+  minItemWidth,
+  maxVisibleCount,
+  minVisibleCount,
+  slideCount,
+  peek,
+  galleryRef
+) {
+  if (!galleryRef.current) {
+    return {visibleCount: 1, maxContainerWidth: Number.MAX_VALUE};
+  }
+  const width = galleryRef.current./* OK */ offsetWidth;
```

Comment:
Hmm. This is a bit of a trouble. It's not a code we can run on SSR side. And also means that we need to listen to resizes. I understand that in SSR mode it should never get this far b/c of the `if` above. But it feels cryptic. Also, might be a source of CLS violations. Any other way to express it? At the very least we could set it up as a state+effect-with-resize like fit-text.

Reply:
I refactored this a bit, could you take another look and clarify concerns if they remain?